src/utils/permission_utils.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/utils/permission_utils.h>
+
+#include <multipass/file_ops.h>
+#include <multipass/platform.h>
+
+namespace mp = multipass;
+namespace fs = mp::fs;
+
+using Path = mp::PermissionUtils::Path;
+
+namespace
+{
+void set_single_permissions(const Path& path, const QFileDevice::Permissions& permissions)
+{
+    QString qpath = QString::fromUtf8(path.u8string());
```

Comment:
There's some funny conversion going on here. `QString` to `u16String` to `u8string` to `QString`.

Reply:
yeah, `QString` -> `u16String` -> `std::filesystem::path` -> `u8string` -> `QString` -> `std::string` -> `char*`. I could probably get it to `QString` -> `u16String` -> `std::filesystem::path` -> `std::string` -> `char*` but any better than that would require some pretty massive changes to everything.